### PR TITLE
Destroy unused handles

### PR DESCRIPTION
**Pull request by amartya4256**
This PR contributes to the cleanup of those handles of Resources which were created for the zoom levels which are not applicable for the existing zoom levels of all the monitor. i.e., if there were 2 monitors with 100% and 200% zoom levels and the seconds monitor is switched to 100% as well. On the DPI_CHANGED event, we trigger a cleanup of handles of the resources using the resource tracker to remove the handles which were created for 200% zoom level since there doesn't exist any monitor with a zoom level of 100% anymore.

With this PR, we also refactor the resource tracker to allow it to track all the resources for win32 all the time but rather using the tracking flag to report errors, since that is the only use case of the tracking flag as of now. By allowing the Resource Tracker to track the resources, we can trigger the cleanup of all the relevant handles from a single point.

Contributes to #62 and #127

### CODE REVIEW

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (fedejeanne) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:852]: Why the change here?

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java
Comment (fedejeanne) [inline bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java:127]: Add some documentation here and mention the fact that the default implementation is empty and therefore sub-classes should override this method if necessary.

**Review by fedejeanne — COMMENTED** [review]
I only skimmed the code, I haven't tested it yet (I'll do it as soon as the PR is ready to be reviewed). I only have 2 minor remarks.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:852]: Its because earlier we had both objects and errors guarded by the lock. But now, object is not guarded by the lock anymore and it would not be null anymore. So, we only wanna return already if both image and errors are null. i think even one check just for errors would also be fine. We don't need to check for objects since the method is called printErrors. Why print only objects in case there's no error? What do you say?

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java:127]: Alright

Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java:136]: How about:

"Destroy all handles of the resource  which are <strong>not</strong> necessary for the given zoom levels. This method is supposed to be overridden by sub-classes that retain handles for different zoom levels."

For the body? 

The `@param` is ok.

Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java:136]: But wer dont know what zoom levels these resources have handles of. So, the only way possible is to tell them to keep these zoom levels and discard others if they have.

Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java:136]: alright

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (fedejeanne) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:957]:
```suggestion
void registerResourceWithZoomSupport(Resource r) {
```

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (fedejeanne) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:61]:
```suggestion
	private Vector<Resource> resourcesWithZoomSupport = new Vector<>();
```

Also, why use `Vector` and not `Collection` / `List`? Do you need synchronization?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: The change has been reverted.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:61]: Yes

Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:61]: Rename?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:957]: Rename?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:61]: Done

**Review by fedejeanne — APPROVED** [review]
The changes look good and produce the expected results. Also, no regressions were found during the testing.

I re-triggered the builds because there were some test failures (for Browser!). If the tests pass in the next run then this PR is good to go on my account.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java:1213]: Please properly format code (here and at other similar places).

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java:1208]: Extracting reused `entry.getKey()` into a self-explaining variable would improve readability.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:61]: Why use outdated Vector instead of, e.g., a synchronized set?

**Review by HeikoKlare — COMMENTED** [review]
I did some sanity checking by testing the change with an SDK product:
- Two monitors, the secondary at 100%, the primary one switched between multiple zoom values
- Once with `swt.autoScale=quarter` and once with default (`integer200`)

I have locally added some logging to see when resources are destroyed and it appears to happen when expected.

I have not conduced an in-depth code review but some minor comments that should be addressed before merging.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: Done

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java]: Done

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java]: Done

Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java]: Fotmatting is still inconsistent.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java]: Pleas also adapt the other places with the same code consistently

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: I meant to use a synchronized set, not to have a set to which every access it synchronized (as that's quite similiar to using a vector).

Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: Oh sorry, its done now

Comment (laeubi) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: I suppose everything has to happen in the UI thread so do we need synchronized (whatever) anyways?
That's actually one of the main points why UI frameworks use "display threads" to prevent the synchronization and possible deadlocks if multiple threads would operate on the same output device.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (laeubi) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:61]: See general comment about `synchronized`, in any case 

```suggestion
	private Set<Resource> resourcesWithZoomSupport  = ConcurrentHashMap.newKeySet();
```

might be a better alternative these days.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (laeubi) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java:361]: init seem to be called everywhere where this code is called, so maybe put it in the init method?!?

Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java:361]: We only wanna register those resources which have multiple handles for different zoom levels. But some other classes like Cursor, Font, etc also call init() but they don't have multiple handles like Image, Path, Pattern, etc so it would be better to call it from the constructors. Also, some constructors like Image(Device device) do not call init() so it will make the registration of such resources inconsistent.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: Valid point. I did not challenge the original answer that synchronization is necessary here (see https://github.com/eclipse-platform/eclipse.platform.swt/pull/1477#discussion_r1783956150), but taking a look at the implementation I do not see any access to the data structure that can happen from outside the UI thread, as both resource instantiation and OS callback processing happen within the UI thread (i.e., the thread of the according `Device`/`Display` instance), don't they?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java]: @amartya4256 I noticed that some of the new code uses spaces instead of tabs to indent. Maybe that's what Heiko means by "still inconsistent"?

Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java]: > I noticed that some of the new code uses spaces instead of tabs to indent. Maybe that's what Heiko means by "still inconsistent"?

Yes, seems to be a mixture of spaces and tabs. E.g., this:
{DF3ACF41-FFEE-4B87-B72C-7560E8E0F39D}

@amartya4256 can you please adapt the inconsistent formatting at all remaining places?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java:361]: Are you satisfied with the answer @laeubi ? May we mark this conversation as "resolved"?

**Comment by fedejeanne** [comment]
Jenkins is still [offline](https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/5131) but the last checks were fine. Merging.

Any additional comments may be addressed in a follow-up PR.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java]: They seem to be done.

I'll mark this as resolved. If there are any open issues, they can be addressed in a follow-up PR.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java
Comment (fedejeanne) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java:361]: @laeubi since we would like to have this in M2 and the current approach LGTM, I merged this PR. 

We can continue this discussion in a follow-up PR if necessary.

**Comment by akurtakov** [comment]
> Jenkins is still [offline](https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/5131) but the last checks were fine. Merging.
> 
> Any additional comments may be addressed in a follow-up PR.

The report you link to is not about Platform and Releng JIPPs so unless you open one specifically for them they will stay offline.

**Comment by HeikoKlare** [comment]
The releng Jenkins has just been restarted.

**Comment by HeikoKlare** [comment]
After Jenkins was back, I just triggered an SWT master build to verify that none of the PRs merged today produced any issues. It finished successfully: https://ci.eclipse.org/releng/job/eclipse.platform.swt/job/master/767/

Regarding reporting of the Jenkins issue, there was a comment that several instances were affected and restarted, so at least I assumed that the infrastructure team checks the whole list of instances:
{CE680D35-5546-4F83-9634-4323ADF91308}

**Pull request merged**
